src/components/PasswordInput/PasswordInput.js

```diff
@@ -0,0 +1,74 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled, { css } from 'react-emotion';
+
+import { InputWrapper } from '../InputWrapper';
+import { Input as StandardInput } from '../Input';
+import { SvgButton } from '../SvgButton';
+import State from '../State/State';
+import RevealIconSvg from './eye.svg';
+import HideIconSvg from './eye-off.svg';
+
+const Input = styled(StandardInput, { label: 'password-input__input' })(
+  ({ theme }) => `
+    padding-right: ${theme.spacings.peta};
+  `
+);
+
+const stylesIcon = css`
+  label: password-input__icon;
+  position: absolute;
+  top: 50%;
+  right: 12px;
+  transform: translateY(-50%);
+`;
+
+const RevealIcon = styled(RevealIconSvg, { label: 'reveal-icon' })(stylesIcon);
+
+const HideIcon = styled(HideIconSvg, { label: 'hide-icon' })(stylesIcon);
+
+/**
+ * PasswordInput component for forms.
+ */
+const PasswordInput = ({ disabled, ...props }) => (
+  <State
+    initialState={false}
+    stateName="isVisible"
+    stateUpdaterName="onToggle"
+    stateUpdater={isVisible => !isVisible}
+  >
```

Comment:
Neat concept - can it only have one state property though? I think that `withState` as a higher order component done like recompose, it's possible to add multiple keys to the state object via composition. Maybe an hoc would be a better choice for that reason?

Reply:
Definitely. So, first you should be able to compose these render prop components (still thinking about that topic). Your render prop would then receive all of the composed states. But, the way the `State` component is implemented, you can pass in any data structure as `initialState` and the `stateUpdater` will receive that structure. In this case, it just so happens that the state is a simple boolean. I haven't fully tested this functionality (should write unit tests for this), but that was the design idea.